src/Workspaces/Remote/ServiceHub/Services/LanguageServer.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.NavigateTo;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    // we need per language server for now since ILanguageClient
+    // doesn't allow multiple content types to be associated with
+    // one language server
+    internal class CSharpLanguageServer : LanguageServer
+    {
+        public CSharpLanguageServer(Stream stream, IServiceProvider serviceProvider)
+            : base(stream, serviceProvider, LanguageNames.CSharp)
+        {
+        }
+    }
+
+    internal class VisualBasicLanguageServer : LanguageServer
+    {
+        public VisualBasicLanguageServer(Stream stream, IServiceProvider serviceProvider)
+            : base(stream, serviceProvider, LanguageNames.VisualBasic)
+        {
+        }
+    }
+
+    internal abstract class LanguageServer : ServiceBase
+    {
+        private static readonly IImmutableSet<string> s_supportedKinds =
+            ImmutableHashSet.Create(
+                NavigateToItemKind.Class,
+                NavigateToItemKind.Constant,
+                NavigateToItemKind.Delegate,
+                NavigateToItemKind.Enum,
+                NavigateToItemKind.EnumItem,
+                NavigateToItemKind.Event,
+                NavigateToItemKind.Field,
+                NavigateToItemKind.Interface,
+                NavigateToItemKind.Method,
+                NavigateToItemKind.Module,
+                NavigateToItemKind.Property,
+                NavigateToItemKind.Structure);
+
+        private readonly string _languageName;
+
+        public LanguageServer(Stream stream, IServiceProvider serviceProvider, string languageName)
+            : base(serviceProvider, stream, SpecializedCollections.EmptyEnumerable<JsonConverter>())
+        {
+            _languageName = languageName;
+
+            StartService();
+        }
+
+        [JsonRpcMethod(Methods.InitializeName)]
+        public object Initialize(int? processId, string rootPath, Uri rootUri, ClientCapabilities capabilities, TraceSetting trace, CancellationToken cancellationToken)
+        {
+            // our LSP server only supports WorkspaceStreamingSymbolProvider capability
+            // for now
+            return new InitializeResult()
+            {
+                Capabilities = new VSServerCapabilities()
```

Comment:
>VSServerCapabilities [](start = 35, length = 20)

What are `VSServerCapabilities`? Seems odd to have VS-specific type referenced in Workspaces layer

Reply:
@gundermanc ?

I guess it has VS since it is VS specific Language Server protocol? and once we move to data contract rather than binary contract used with client programming model. our layering is a bit weird I guess. since we can implement those at the lower level and we don't actually care whether the other side is VS or not as long as they understand VS protocol or we can implement the protocol without VS dependencies.